### PR TITLE
PUBDEV-4056: ARFF parser, fix bug triggered by spaces in attribute names

### PR DESCRIPTION
**Pull request by michalkurka**

### CODE REVIEW

File: h2o-core/src/main/java/water/parser/ARFFParser.java [hunk 1]
```diff
@@ -127,55 +127,63 @@ private static int readArffHeader(int offset, ArrayList<String> header, byte[] b
...
+          for (int j = 0; j < domains[i].length; j++)
+            domains[i][j] = domains[i][j].trim();
+          if (domains[i][0].length() > 0)
+            ctypes[i] = Vec.T_CAT; // case of {A,B,C} (valid list of factors)
         }
-        // only get here if data is invalid ARFF
-        throw new ParseDataset.H2OParseException("Unexpected line.");
+
+        if (ctypes[i] == Vec.T_BAD)
+          throw new ParseDataset.H2OParseException("Unexpected line.");
```
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ARFFParser.java]: Makes sense to have a more meaningful error message: col number, maybe col name, and more detailed info.

File: h2o-core/src/main/java/water/parser/ARFFParser.java [hunk 2]
```diff
@@ -127,55 +127,63 @@ private static int readArffHeader(int offset, ArrayList<String> header, byte[] b
...
+        int sepIdx = spec.lastIndexOf(' ');
+        if (sepIdx < 0) {
           throw new ParseDataset.H2OParseException("Expected @ATTRIBUTE to be followed by <attribute-name> <datatype>");
         }
-        labels[i] = data[i][1];
-        String type = data[i][2];
+        final String type = spec.substring(sepIdx + 1).trim();
         domains[i] = null;
+        ctypes[i] = Vec.T_BAD;
         if (type.equalsIgnoreCase("NUMERIC") || type.equalsIgnoreCase("REAL") || type.equalsIgnoreCase("INTEGER") || type.equalsIgnoreCase("INT")) {
```
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ARFFParser.java]: (Almost) all these if/assign can be replaced with a static map, string -> type, leaving only operations with '}'

File: h2o-core/src/main/java/water/parser/ARFFParser.java [hunk 3]
```diff
@@ -127,55 +127,63 @@ private static int readArffHeader(int offset, ArrayList<String> header, byte[] b
...
-          for(int j = 2; j < data[i].length; j++) {
-            builder.append(data[i][j]);
-          }
-          domains[i] = builder.toString().replaceAll("[{}]", "").split(",");
-          if (domains[i][0].length() > 0) {
-            // case of {A,B,C} (valid list of factors)
-            ctypes[i] = Vec.T_CAT;
-            continue;
-          }
+        else if (type.endsWith("}")) {
```
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ARFFParser.java]: I got a feeling that a regex would make more sense here.

File: h2o-core/src/test/java/water/parser/ARFFParserTest.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package water.parser;
+
+import org.junit.Test;
+import water.fvec.Vec;
+
+import static org.junit.Assert.*;
+
+/**
+ * Unit test for methods of ARFFParser, for integration tests {@see ParserTestARFF}
```
Comment (vpatryshev) [inline h2o-core/src/test/java/water/parser/ARFFParserTest.java]: Good stuff, but I'd add a couple of negative tests.

**Comment by vpatryshev** [comment]
my comments are inline

File: h2o-core/src/main/java/water/parser/ARFFParser.java (hunk 2 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ARFFParser.java]: There are two special cases: categorical and relational type (which we do not support and produce an exception). I think this justifies the use of the IF statements - in my opinion.

File: h2o-core/src/main/java/water/parser/ARFFParser.java (hunk 3 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ARFFParser.java]: Regex would work too, what I need to find is the occurrence of pattern "{[^{]*}$" in the type specification. I can rewrite it with regex. Do you think it is worth it (do you see an issue with the current code)?

File: h2o-core/src/main/java/water/parser/ARFFParser.java (hunk 1 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ARFFParser.java]: Absolutely, I will try to improve it.

File: h2o-core/src/test/java/water/parser/ARFFParserTest.java (hunk 4 above)
Comment (michalkurka) [inline h2o-core/src/test/java/water/parser/ARFFParserTest.java]: Good suggestion, thank you

Comment (michalkurka) [inline h2o-core/src/test/java/water/parser/ARFFParserTest.java]: DONE

Comment (michalkurka) [inline h2o-core/src/test/java/water/parser/ARFFParserTest.java]: DONE

**Comment by michalkurka** [comment]
@h2o-ops please review

**Comment by michalkurka** [comment]
@h2o-ops please test

**Pull request merged**